src/services/codefixes/convertToAsyncFunction.ts

```diff
@@ -0,0 +1,471 @@
+    }
+
+    function replaceNodes(changes: textChanges.ChangeTracker, sourceFile: SourceFile, oldNodes: Node[], allNewNodes: Map<Node[]>) {
+        for (const statement of oldNodes) {
+            const newNodes = allNewNodes.get(getNodeId(statement).toString());
+            if (newNodes) {
+                changes.replaceNodeWithNodes(sourceFile, statement, newNodes);
+            }
+        }
+    }
+
+    function getConstIdentifiers(synthNamesMap: Map<SynthIdentifier>): Identifier[] {
+        const constIdentifiers: Identifier[] = [];
+        synthNamesMap.forEach((val) => {
+            if (val.numberOfUsesOriginal === 0) {
+                constIdentifiers.push(val.identifier);
+            }
+        });
+        return constIdentifiers;
+    }
+
+    function findLastDotThens(func: FunctionLikeDeclaration, checker: TypeChecker): Map<boolean> {
+        if (!func.body) {
+            return createMap();
+        }
+
+        // maps nodes to boolean - true indicates that there is another .then() in the callback chain
+        const lastDotThen: Map<boolean> = createMap();
+
+        forEachChild(func.body, function visit(node: Node) {
+            const nodeType = checker.getTypeAtLocation(node);
+            if (isCallExpression(node) && nodeType && !!checker.getPromisedTypeOfPromise(nodeType) && hasPropertyAccessExpressionWithName(node, "then")) {
+                // false - there is no following .then() in the callback chain
+                lastDotThen.set(getNodeId(node).toString(), false);
+
+                forEachChild(node, function checkChildren(child: Node) {
+                    forEachChild(child, checkChildren);
+
+                    const nodeType = checker.getTypeAtLocation(child);
+                    if (isCallExpression(child) && nodeType && !!checker.getPromisedTypeOfPromise(nodeType) && hasPropertyAccessExpressionWithName(child, "then")
+                        && child.parent && !isPropertyAccessExpression(child.parent) && lastDotThen.get(getNodeId(child.parent).toString()) === false) {
+
+                        // false - there is no following .then() in the callback chain
+                        lastDotThen.set(getNodeId(child).toString(), false);
+                    }
+                    else if (isCallExpression(child) || isIdentifier(child) && nodeType && checker.getPromisedTypeOfPromise(nodeType)) {
+
+                        // true - there is a following .then() in the callback chain
+                        lastDotThen.set(getNodeId(child).toString(), true);
+                    }
+                });
+            }
+            else {
+                forEachChild(node, visit);
+            }
+        });
+
+        return lastDotThen;
+    }
+
+    function isFunctionRef(node: Node): boolean {
+        const callExpr = climbPastPropertyAccess(node);
+        return !isCallExpression(callExpr) || callExpr.expression !== node;

```

Comment:
I'm not sure I understand what this function is testing for, but it seems like `callExpr.expression !== node` will be true whenever it is reached.

Reply:
(Because if `callExpr` is a call expression, then it isn't a property access expression so `climbPastPropertyAccess` must have returned `node`, which does not have itself as a child.)

Reply:
It looks like the parameter could be `node: Identifier`.